impeller/aiks/paint_pass_delegate.cc

```diff
@@ -44,4 +48,90 @@ std::shared_ptr<Contents> PaintPassDelegate::CreateContentsForSubpassTarget(
                                             effect_transform);
 }
 
+/// OpacityPeepholePassDelegate
+/// ----------------------------------------------
+
+OpacityPeepholePassDelegate::OpacityPeepholePassDelegate(
+    Paint paint,
+    std::optional<Rect> coverage)
+    : paint_(std::move(paint)), coverage_(coverage) {}
+
+// |EntityPassDelgate|
+OpacityPeepholePassDelegate::~OpacityPeepholePassDelegate() = default;
+
+// |EntityPassDelgate|
+std::optional<Rect> OpacityPeepholePassDelegate::GetCoverageRect() {
+  return coverage_;
+}
+
+// |EntityPassDelgate|
+bool OpacityPeepholePassDelegate::CanElide() {
+  return paint_.blend_mode == BlendMode::kDestination;
+}
+
+// |EntityPassDelgate|
+bool OpacityPeepholePassDelegate::CanCollapseIntoParentPass(
+    EntityPass* entity_pass) {
+  // Note: determing whether any coverage intersects has quadradic complexity in
+  // the number of rectangles, and depending on whether or not we cache at
+  // different levels of the entity tree may end up cubic. In the interest of
+  // proving whether or not this optimization is valuable, we only consider very
+  // simple peephole optimizations here - where there is a single drawing
+  // command wrapped in save layer. This would indicate something like an
+  // Opacity or FadeTransition wrapping a very simple widget, like in the
+  // CupertinoPicker.
+  if (entity_pass->GetEntityCount() > 3) {
```

Comment:
One concern I have with working on this optimization now: subpasses are extra expensive right now on the metal backend due to the whole waitForScheduled, and also potentially due to lack of multithreaded encoding.

We might tune these thresholds too high and end up trading a bunch of raster time for decreasing returns of GPU utilization.

Reply:
Yes! We're terribly underutilizing the GPU here. Sibling EntityPasses are fundamentally independent and can be evaluated in parallel on the CPU (not a huge impact) and executed in parallel on the GPU (absolutely enormous impact). Once we have a proper [synchronization solution](https://github.com/flutter/flutter/issues/120406), we can remove all of the host<->device syncs that are blocking the CPU and saturate the GPU with parallelizable work.

Reply:
Although I think we added texture atlas sharing between EntityPasses at some point. That (and any other dependencies shared between EntityPasses) need to be reverted, of course.

Reply:
Can we get around that by making glyph atlas construction happen eagerly? Otherwise, glyph atlas construction and upload takes a substantial amount of time.

Reply:
or, more specifically, can we push construction of the glyph atlas including all text frames into a different thread and then use whatever resource tracking mechanism we create to ensure the subsequent commands that depend on this are ordered correctly?

Reply:
> Can we get around that by making glyph atlas construction happen eagerly

Yeah I think that's probably the best option -- just render the whole thing at once and upload everything just before the EntityPass run. Then the uploaded atlases become read only for the frame and don't restrict parallelism.